Feat: Adds Card Component

ataideverton opened the pull request:
## What's the purpose of this pull request?

- To  add the card component. The card component is integral part of the B2B Self Management design but it's also a generic component and according to the design team, will be used on things like MyAccount in the future.

![image](https://github.com/user-attachments/assets/897c55b5-63be-40c6-a8ec-51153c098d81)

[Figma](https://www.figma.com/proto/r3swf1BAZB3yJa9FrgcyXx/Buyer-Organizations-%C2%B7-2024?node-id=128-37169&t=SCMy7ME6H9mqtsiN-0&scaling=min-zoom&page-id=128%3A31910)

## How to test it?

### Starters Deploy Preview

[Preview](https://sfj-8b4430d--b2bfaststoredev.preview.vtex.app/self-management)
[PR on Starter](https://github.com/vtex-sites/b2bfaststoredev.store/pull/88)

<!--- Add a link to a deploy preview from `gatsby.store` AND `nextjs.store` with this branch being used. --->

<!--- Tip: You can get an installable version of this branch from the CodeSandbox generated when this PR is created. --->

renatamottam commented on packages/ui/src/components/molecules/Card/styles.scss:
A couple feedbacks about this styling file:
- always list all the available tokens on the beginning of the file, you can add a title if needed, example:
https://github.com/vtex/faststore/blob/32c36b739087214e8abb6a78da4623033143ec97/packages/ui/src/components/molecules/Alert/styles.scss#L22
- we usually don't create `border-style` if it's `solid`
- try to use the semantic colors available and avoid creating new colors only for the component
- you can use `fs-spacing` tokens for paddings and margins
- we have `font-weight` tokens already
- to improve the file's readability, add spaces between token name and value, example:
https://github.com/vtex/faststore/blob/32c36b739087214e8abb6a78da4623033143ec97/packages/ui/src/components/molecules/Alert/styles.scss#L7-L9

ataideverton replied:
Can you check again and see if this new version is ok?

renatamottam commented on packages/components/src/molecules/Card/Card.tsx:
```diff
@@ -0,0 +1,61 @@
+
+export interface CardProps extends Omit<HTMLAttributes<HTMLDivElement>, 'role'> {
+    /**
+    * Specfies the text to be loaded into the header.
```
```suggestion
    * Specifies the text to be loaded into the header.
```

renatamottam commented on packages/components/src/molecules/Card/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default } from './Card'
+export type { CardProps } from './Card'
```
```suggestion
export type { CardProps } from './Card'

```

hellofanny commented on packages/components/src/molecules/Card/Card.tsx:
```diff
@@ -0,0 +1,61 @@
+}, ref) {
+    return (
+        <div data-testid={testId} ref={ref} data-fs-card style={{ maxWidth: `${maxWidth}` }}>
+            <div data-fs-card-header style={{ display: 'flex', justifyContent: 'space-between' }}>
```
`style={{ display: 'flex', justifyContent: 'space-between' }}>` 

**todo:** Can you please move this inline style to the [/molecules/Card/styles.scss](https://github.com/vtex/faststore/pull/2398/files#diff-ecc6db095738567755bce1001bfebefdfe9daa6031016b2456764b743cde2f63) stylesheet? 

We usually don't use inline css.

hellofanny commented on packages/components/src/molecules/Card/Card.tsx:
```diff
@@ -0,0 +1,61 @@
+}
+
+const Card = forwardRef<HTMLDivElement, CardProps>(function Card({
+    title,
+    children,
+    maxWidth,
+    testId = 'fs-card',
+    iconName,
+    iconColor,
+    iconAction
```
```suggestion
    title,
    maxWidth,
    testId = 'fs-card',
    iconName,
    iconColor,
    iconAction,
    children,
    ...otherProps
```

hellofanny replied:
**todo**: Adding otherProps to allow additional props to be passed if needed (such as styles...)

hellofanny commented on packages/components/src/molecules/Card/Card.tsx:
```diff
@@ -0,0 +1,61 @@
+    iconAction
+}, ref) {
+    return (
+        <div data-testid={testId} ref={ref} data-fs-card style={{ maxWidth: `${maxWidth}` }}>
```
```suggestion
        <div data-testid={testId} ref={ref} data-fs-card style={{ maxWidth: `${maxWidth}` }}  {...otherProps} >
```

hellofanny commented on packages/components/src/molecules/Card/Card.tsx:
```diff
@@ -0,0 +1,61 @@
+    iconAction
+}, ref) {
+    return (
+        <div data-testid={testId} ref={ref} data-fs-card style={{ maxWidth: `${maxWidth}` }}>
```
**suggestion:**  How about using <section> instead of <div> for this wrapper?  Since represents a distinct section of content would be a good choice.

ataideverton replied:
I think it's a good idea, updated it

hellofanny reviewed (changes requested):
Hellooo! Great job!
I've partially reviewed it and added a few suggestions. Can you take a look, please? 🚀

ataideverton commented:
> Hellooo! Great job! I've partially reviewed it and added a few suggestions. Can you take a look, please? 🚀

Implemented all the suggestions, if you can review it again

hellofanny commented on packages/components/src/molecules/Card/Card.tsx:
```diff
@@ -0,0 +1,58 @@
+  /**
+  * Specifies the text to be loaded into the header.
+  */
+  title?: string
```
**question:** Shouldn't `title` be required? (If not, I think we should remove the fs-card-title div)

ataideverton replied:
It should, yes

hellofanny commented on packages/components/src/molecules/Card/Card.tsx:
```diff
@@ -0,0 +1,58 @@
+      <div data-fs-card-header>
+        <div data-fs-card-title>{title}</div>
+        {iconName && (
+          <Icon name={iconName} width={20} height={20} weight="bold" color={iconColor} onClick={iconAction} />
```
We usually have a data-attr for icon [customization](https://github.com/vtex/faststore/blob/ff12ccfdf1cc909dc708912fa5b0c95e6aaaef57/packages/ui/src/components/molecules/Gift/styles.scss#L24), for example `data-fs-card-icon` and  a `--fs-card-icon-color` token to change the icon color. 

cc: @renatamottam

ataideverton replied:
Changed it

hellofanny reviewed (commented):
Since, the provided preview link was not working, I tried this component in the [starter](https://github.com/vtex-sites/starter.store/pull/512). I think it's working as expected. 

<img width="337" alt="image" src="https://github.com/user-attachments/assets/6cfa1f35-c578-480d-89b7-03706de022d1">

@renatamottam  Do you think we need to test anything else?

ataideverton commented on packages/components/src/molecules/Card/Card.tsx:
```diff
@@ -0,0 +1,58 @@
+      <div data-fs-card-header>
+        <div data-fs-card-title>{title}</div>
+        {iconName && (
+          <Icon name={iconName} width={20} height={20} weight="bold" color={iconColor} onClick={iconAction} />
```
I think it could, yeah

I will do some tests and change it later today, thanks

hellofanny commented on packages/ui/src/components/molecules/Card/styles.scss:
```diff
@@ -0,0 +1,45 @@
+    background-color: var(--fs-card-header-bkg-color);
+    font-weight: var(--fs-card-header-font-weight);
+    display: flex;
+    justify-content: space-between;
+    [data-fs-button]{
+      --fs-button-height: var(--fs-spacing-4);
+    }
+    [data-fs-icon]{
+      color: var(--fs-card-header-icon-color);
+    }
+  }
```
```suggestion
    justify-content: space-between;
    
    [data-fs-button]{
      --fs-button-height: var(--fs-spacing-4);
    }
    
    [data-fs-icon]{
      color: var(--fs-card-header-icon-color);
    }
  }
```

hellofanny commented:
@ataideverton Thank youuu! It looks great with the IconButton! LGTM!

<img width="300" alt="image" src="https://github.com/user-attachments/assets/ba3566f1-9741-49c4-a02a-c3ac30b79381">


Regarding the docs updated with this new component, let's open a request so @Mariana-Caetano can handle it :)

ataideverton commented:
> @ataideverton Thank youuu! It looks great with the IconButton! LGTM!
> 
> <img alt="image" width="300" src="https://private-user-images.githubusercontent.com/3356699/361043166-ba3566f1-9741-49c4-a02a-c3ac30b79381.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2NzI2ODYsIm5iZiI6MTcyNDY3MjM4NiwicGF0aCI6Ii8zMzU2Njk5LzM2MTA0MzE2Ni1iYTM1NjZmMS05NzQxLTQ5YzQtYTAyYS1jM2FjMzBiNzkzODEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjZUMTEzOTQ2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDkyNzU1NGVlNzhlNjdkNzRlYjhhMDg5MTQ2YmIyNTFjNDFjMmJhMTNlNTFmYTdmYmRjYTMwMGE0NWNiOWE3OSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.nOYfSrOOtKsefEJfcgFNeK8JCPNxfCJVzkJvWco5tG4">
> Regarding the docs updated with this new component, let's open a request so @Mariana-Caetano can handle it :)

I did, the docs update is just so she can have a base to work from

renatamottam commented:
@ataideverton could you provide some images showing how the card is after the reviews (with brandless and with vtex way theme)?
